docs/source/traits_user_manual/advanced.rst

```diff
@@ -38,15 +38,23 @@ attribute, with a name based on the name of the trait attribute:
 
 .. method:: _name_default()
 
-This method initializes the *name* trait attribute, returning its initial value.
-The method overrides any default value specified in the trait definition. The
-initializer is called when:
+This method returns the default value for the *name* trait attribute and it
+overrides any default value specified in the trait definition.
+
+Similar to static default values, default values defined dynamically are
+considered to have existed **prior to** setting object state during
+initialization. For performance purposes, a default initializer is called when:
 
 1. the attribute value is accessed the first time or
 2. an instance is constructed with a specific value, if there is a change
    handler defined for the trait. This is needed so the default can be reported
    as the old value (see :ref:`static-notification`).
 
+Without any trait change notification definitions, it is possible to use
+default initializers to initialize attributes using object state after
+instantiation. However such usage will often lead to unexpected behavior when
+notifications for trait changes are defined.
```

Comment:
This is a bit vague and ominous.  Something like:

> It is possible to use default initializers to defer expensive initialization of attribute values, but this relies on not having any trait notifiers on the trait.  This can be difficult in practice, since trait notifiers can be set up by any object, and the creation of these notifiers is frequently done automatically as part of things like Property traits, the delegation infrastructure, complex dynamic trait observation, or the creation of a TraitsUI GUI.  This can cause the expensive initialization to occur at unexpected times, and in a some cases may actually result in worse overall performance.


I think that covers the unexpected behaviour more clearly, but there may be other cases than the "defer expensive initialization" that we care about.

Reply:
I see what you mean - I agree the original sentence is not very specific on the details and therefore looks overly worrying.  I tried to rework it using some of your recommendations here. Please take a look again.

From what I have seen, default initializers in projects that would run into this problem tend to be small setup code: they are not expensive calculations at all. They are done as an alternative to `traits_init` or overriding `__init__`. The laziness nature was an added bonus (when it works). The reasons why those default initializers can be sources of surprises when they depend on post-initialization states, which works until notifications bring the evaluation of default forward.

Reply:
(I also avoided the word "Notifier", it is somewhat an insider term that I am not sure the end users are familiar with)